[9주차] Kafka 활용 서비스 개선

y-00jin opened the pull request:
### **커밋 설명**
- Kafka 개념 문서 작성 — #50  
- 결제 성공 이벤트 Kafka 메시지로 발행 — #51 
- 도메인 중심으로 프로젝트 구조 리팩토링 — #52

y-00jin committed:
REFACTOR: 프로젝트 구조 리팩토링 - 공통 config (#52)

- 공통 config 파일 정리

y-00jin committed:
REFACTOR: 프로젝트 구조 리팩토링 - concert (#52)

y-00jin committed:
REFACTOR: 프로젝트 구조 리팩토링 - user (#52)

y-00jin committed:
REFACTOR: 프로젝트 구조 리팩토링 - 대기열 queue (#52)

y-00jin committed:
REFACTOR: 프로젝트 구조 리팩토링 - reservation (#52)

y-00jin committed:
REFACTOR: 프로젝트 구조 리팩토링 - payment (#52)

y-00jin committed:
REFACTOR: 프로젝트 구조 리팩토링 - lock (#52)

y-00jin committed:
REFACTOR: mock API 컴포넌트 infrastructure로 이동(#52)

y-00jin committed:
REFACTOR: 구조 리팩토링으로 인한 Test코드 수정(#52)

y-00jin committed:
FEAT: 결제 성공 이벤트 Kafka 메시지로 발행하도록 DataPlatformClient 리팩토링 (#51)

- 기존 log 기반 Mock API 호출 방식에서 Kafka 메시지 발행 방식으로 변경
- 결제 성공시 payment-success 토픽에 PaymentSuccessEvent 발행

y-00jin committed:
DOCS: Kafka 개념 문서 작성 (#50)

Isaac482763 commented on src/main/java/kr/hhplus/be/server/common/config/kafka/KafkaConfig.java:
```diff
@@ -0,0 +1,59 @@
+        configProps.put(ConsumerConfig.GROUP_ID_CONFIG, paymentGroup); // Consumer 그룹 ID
+        configProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        configProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, JsonDeserializer.class);
+        configProps.put(JsonDeserializer.TRUSTED_PACKAGES, "*");
```
*로 설정하면 모든 패키지의 클래스를 역직렬화할 수 있습니다.
보안을 고려하여 필요한 특정 패키지만 지정하여 사용하는것이 어떨까합니다.

예시
```
configProps.put(JsonDeserializer.TRUSTED_PACKAGES, "kr.hhplus.be.server.payment.domain.event");
```

Isaac482763 commented on src/main/java/kr/hhplus/be/server/dataPlatform/infrastructure/persistence/DataPlatformClient.java:
```diff
@@ -0,0 +1,22 @@
+
+    private final KafkaTemplate<String, Object> kafkaTemplate;
+
+    public void sendPaymentInfo(PaymentSuccessEvent event) {
+        kafkaTemplate.send(KafkaTopicConstants.PAYMENT_SUCCESS, event.getPaymentId().toString(), event);
+        log.info("데이터플랫폼 Mock API 호출: {}", event.getPaymentId());
+    }
```
네트워크 장애 등으로 메시지가 유실되는 경우를 방지하기 위해서.

Producer에서는 at-least-once. 적어도 한번은 메시지가 발송되도록 보장해야합니다.
```
public void sendPaymentInfo(PaymentSuccessEvent event) {
        ListenableFuture<SendResult<String, Object>> future = kafkaTemplate.send(
            KafkaTopicConstants.PAYMENT_SUCCESS,
            event.getPaymentId().toString(),
            event
        );

        future.addCallback(new ListenableFutureCallback<>() {
            @Override
            public void onSuccess(SendResult<String, Object> result) {
                log.info("데이터플랫폼 전송 성공: paymentId={}, offset={}",
                    event.getPaymentId(),
                    result.getRecordMetadata().offset()
                );
            }

            @Override
            public void onFailure(Throwable ex) {
                log.error("데이터플랫폼 전송 실패: paymentId={}, 원인={}",
                    event.getPaymentId(),
                    ex.getMessage(), ex
                );
                // 재시도 로직 추가.
            }
        });
    }
```

Isaac482763 commented:
안녕하세요 유진님.
이번주도 꼼꼼하게 잘 해주셨네요 ㅎㅎ

카프카 producer를 잘 추가해 주셨는데요.
위에 리뷰 내용처럼 재시도 로직만 추가하면 훨씬 좋을거같아요.

그리고 지금 코드에서  consumer코드는 안보이는데요.
컨수머쪽도 추가해 보시면 좋을거같습니다.

producer에서 메시지가 유실되는 경우가 있듯이.
consumer에서는 동일 메시지가 여러번 수신되는 경우가 발생할수도 있습니다.
이런 경우 동일한 메시지를 여러번 처리하더라도 결과가 동일하도록 보장하는것을 멱등성 이라고 하는데요.
이거를 고려해서 한번 처리한 메시지는 기록해놓고 매번 메시지를 받았을때 이미 처리한것인지 확인하는 절차가 필요합니다.
이런식으로 컨수머쪽도 한번 작성해보시면 좋을것같습니다.!

y-00jin committed:
REFACTOR: KafkaConfig에서 역직렬화 신뢰 패키지 명시 (#51)

y-00jin committed:
REFACTOR: Kafka 메시지 전송 실패 시 딜레이 재시도 로직 추가 (#51)

y-00jin committed:
REFACTOR: 결제 성공 Consumer 추가 및 Producer 클래스 네이밍 개선 (#51)

- 결제 성공 Consumer(PaymentSuccessConsumer) 추가
- DataPlatformClient → PaymentProducer로 이름 변경

y-00jin committed:
REFACTOR: 구조 리팩토링으로 인한 Test코드 수정(#52)

y-00jin committed:
TEST: 결제 성공 Kafka Producer/Consumer 통합 및 단위 테스트 추가 (#51)
